### DISCUSSION TITLE
Determining which channel is being used

### DISCUSSION THREAD

**Discussion by s-clerc** [joaotavora/sly#439] [Ideas]
I was thinking about implementing inline media support for Slyblime. The client side support will be fairly easy (at least for Sublime Text) but based on [this proof of concept for SLIME](https://readevalprint.tumblr.com/post/617089509006491648/previewing-images-in-and-out-of-your-slime-repl) I assume a Emacs contrib wouldn't be that difficult either. The main difficulty with the approach there is that the Common Lisp code needs a way to figure out which channel it is outputting to, so that it can send the `write-media` command, so I was wondering if that were possible as I've not been able to figure that out while reading the code.

(An alternative approach, used by Geiser, is just to print `#<Image: ...>` and then perform a replacement, but that approach seems messier than the concept above)

**Comment by joaotavora** [comment]
Interesting.  I'm fine with adding this, but maybe it's already there.  How familiar are you with the `channel` and `listener` classes in `slynk-mrepl.lisp`?  What do you think is missing?

**Comment by s-clerc** [comment]
I've had a look at both classes, but in general I'm more familiar with `channel`. I think the main change needed is to do with `mrepl-eval-1`, where special-variable `*channel*` or similar is introduced into the evaluation environment, which points to the current channel.

Then we could write something along the lines of
```lisp
(defun write-image (path &key allow-outside-channel)
  (unless slynk::*channel* (error "You must use this method in a channel"))
  (with-open-file (in path :direction :input  :element-type '(unsigned-byte 8))
    (let* ((arr (make-array (file-length in) :element-type '(unsigned-byte 8)))
           (b64 (progn (read-sequence arr in) 
                       (cl-base64:usb8-array-to-base64-string arr))))
      (send-to-remote-channel
          slynk:*channel*
          `(:write-image ((:data ,b64 :type (file-type path))) path)))))
```
